### ISSUE TITLE
File is not closed after `torch::from_file(shared=False)` is called

### ISSUE THREAD

**Issue by peterjc123** [pytorch/pytorch#45962]
## 🐛 Bug

`torch::from_file` uses `THMapAllocator` and `THMapAllocator` binds to a file mapping. On Windows, until the file mapping is closed, you could not delete it. This makes the following case problematic.
```c++
char filename[MAX_PATH] = "some temp paths";
auto data =
      torch::from_file(filename, /*shared=*/false, /*size=*/size, torch::kU8);
// Here we expect the file should be closed.
// But actually it is not closed
// It has to be explicitly closed here so that the following clause won't fail.
remove(filename); // This will fail.
```

## Possbile workgrounds
1. Does a copy and swap so that the original allocator is closed.
2. Don't use file mapping for that case. Instead, we read all data into memory and close the file. (An additional option?)
3. We may provide a context manager for releasing the tensor. e.g. in Python
```py
with tempfile.NamedTemporaryFile() as f:
    with torch.TensorAutoRemoveContext():
         t = torch.from_file(f, shared=False, size=size, dtype=torch.uint8)
    # When we exited from the context, we explicitly delete the tensors created within the context
    # In other words, this is equivalent to `del t` here.
```

Any thoughts. @ezyang @malfet ?

**Comment by ezyang** [comment]
If you want to copy it to memory, you can always do a `clone()` after the `from_file()` and then arrange for the original Tensor to go out of scope (thus closing the handle).

**Comment by peterjc123** [comment]
@ezyang Yes, that's what we are trying to do to fix the problem.
